### ISSUE TITLE
Impossible to open an encrypted database with version 1.8.1

### ISSUE THREAD

**Issue by NeoXTof** [moneymanagerex/moneymanagerex#6971]
MMEX version:
 - [ ] 1.7.0
 - [ ] 1.8.0
 - [X] 1.8.1
 - [ ] 1.8.2
 - [ ] Other (please specify)

Note: bug reporters are expected to have verified the bug still exists
either in the last stable version of MMEX or on updated development code
(master branch).

Operating System:
 - [x] Windows 64 (windows 10 pro 64bits 22H2)
 - [x] Mac OSX
 - [x] Linux 

# Description of the bug

When trying to open an encrypted database putting the correct password, I have this error message and the database is not opening . I reinstalled version 1.8.0 and the database is opened with the same password properly

![image](https://github.com/user-attachments/assets/c7a53db2-80a4-4c31-8f7d-bf68f6bb7d7c)

# Reproduction

Is the bug reproducible? 
 - [x] Always 
 - [ ] Randomly 
 - [ ] Happened only once

Reproduction steps:

1. launch MoneyManagerEX 1.8.1
2. input the database password
3. error message appear

Expected result:
Access to encrypted database

Actual result:
I can not access my encrypted database that is opening properly with v1.8.0

# Additional information

If you have a backtrace for a crash or a warning, paste it here.

**Comment by cmbarbu** [comment]
I am encountering the same issue.

I thought it might be possible to save an unencrypted DB in 1.8.0 and encrypt it again in 1.8.1. However, attempting to encrypt the DB in 1.8.1 fails with the following error message:

`SQL logic error[1]: Rekeying failed. Pagesize cannot be changed for an encrypted database.`

![sql-logic-error](https://github.com/user-attachments/assets/56638b46-ef0d-4618-a8aa-d6213066f6cd)

**Comment by Astermast3r** [comment]
Same problem to me, I upgraded from 1.6.1 release on Windows11, but I receive a message of incorrect password or unrecognized MMEX database, even if I try with the utility to decrypt the encrypted database, it is not recognized. In the past I had problems updating the quotes and I decided not to update the program anymore, now I made the mistake of doing so and I can no longer access my Database! I also tried to go back to the previous version, 1.6.1, but it gives me an error of database version 1.8 not recognized. How to solve? Is there a tool to reconvert the old encrypted DB?
![moneymanager_encrypted_DB_error](https://github.com/user-attachments/assets/1ec8bb07-6c11-4690-ab13-9180452e82c7)

**Comment by WinterBowler** [comment]
Also having the same problem. emb installed on Win 10 Pro. Reinstalled ver 1.8.0 and database opened Ok

**Comment by Astermast3r** [comment]
OK I also solved intalling ver. 1.8.0, thanks WinterBowler!

**Comment by i-embree** [comment]
Ditto here on the title issue - unable to open my .emb file on v1.8.1, but okay on v1.8.0 - running on Win11

**Comment by n-stein** [comment]
The encryption cipher for 1.8.1 was upgraded from AES-128 to AES-256. To upgrade your DB to be compatible with the new cipher:

1. Download wxSQLitePlus
2. Open your desktop database file (either encrypted or plain)
3. File -> Encryption Key will upgrade the cipher to SQLCipher/AES256 in place.

Instructions copied from [this forum post](https://forum.moneymanagerex.org/viewtopic.php?p=24957#p24957) by @guanlisheng.

**Comment by guanlisheng** [comment]
@reddyshyam, which version of wxSQLitePlus you used?

you can also use v1.8.0 to convert it as MMB as another workaround (mentioned in the forum [thread](https://forum.moneymanagerex.org/viewtopic.php?p=25564#p25564))
```
- save your 1.8.0 database as a .mmb
- open that with 1.8.1
- save it again as a .emb
```

**Comment by valsamis-d** [comment]
> @reddyshyam, which version of wxSQLitePlus you used?
> 
> you can also use v1.8.0 to convert it as MMB as another workaround (mentioned in the forum [thread](https://forum.moneymanagerex.org/viewtopic.php?p=25564#p25564))
> 
> ```
> - save your 1.8.0 database as a .mmb
> - open that with 1.8.1
> - save it again as a .emb
> ```

@guanlisheng unfortunately this does not work. It fails with the error message stated in:

https://github.com/moneymanagerex/moneymanagerex/issues/6971#issuecomment-2508654018

**Comment by reddyshyam** [comment]
> @reddyshyam, which version of wxSQLitePlus you used?
> 
> you can also use v1.8.0 to convert it as MMB as another workaround (mentioned in the forum [thread](https://forum.moneymanagerex.org/viewtopic.php?p=25564#p25564))
> 
> ```
> - save your 1.8.0 database as a .mmb
> - open that with 1.8.1
> - save it again as a .emb
> ```

https://github.com/moneymanagerex/moneymanagerex/issues/6971#issuecomment-2508893276

**Comment by valsamis-d** [comment]
@reddyshyam followed the steps below:

```
- save your 1.8.0 database as a .mmb
- open that with 1.8.1
- save it again as a .emb
```

but when trying to save as .emb I'm getting the following error:

![image](https://github.com/user-attachments/assets/c6fd3eaa-cd48-4900-8ca3-92fafbf4c1aa)

I also tried the following in wxSQLitePlus v0.3.8:

```
1. Download [wxSQLitePlus](https://github.com/guanlisheng/wxsqliteplus/releases)
2. Open your desktop database file (either encrypted or plain)
3. File -> Encryption Key will upgrade the cipher to SQLCipher/AES256 in place.
```

In Step #2 I'm getting the following warning:

![image](https://github.com/user-attachments/assets/15e5a5a6-c6b1-4b00-b3fe-b2efca3ee69b)

After clicking OK, wxSQLitePlus prompts for a password and after entering it the database opens:

![image](https://github.com/user-attachments/assets/115c9106-0edd-4ada-a4ab-3fc958ae6e70)

Then in Step #3, I'm getting the following error:

![image](https://github.com/user-attachments/assets/f6b4a76a-fcef-4985-9921-1bcb1d53802a)

**Comment by grumpy235** [comment]
The following worked for me when moving from 1.8.0 with an emb to 1.8.1 on Win 11

1.  Backup your encrypted database which is working on 1.8.0 (just to be sure!)
2.  Open 1.8.0 and open a new database just to ensure your live database is not open, probably over-kill
3.  Go to "Tools/Database/Decrypt Encrypted Database"
4.  Specify your 1.8.0 encrypted database name and save with the same name **but with the 'mmb' suffix** - Mydb.emb becomes Mydb.mmb.
5.  Close 1.8.0 and rename Mydb.emb to Mydb180.emb as you will create Mydb.emb again later
7.  Open 1.8.1 and open your decrypted database Mydb.mmb - it should display
8.  Go to "File/Sava Database as.."
9.  At the prompt specify Mydb for the file name and the type as Encrypted MMEX Database (*.emb), hit enter
10. When prompted for the password then enter what you want, I used the same as on my 1.8.0 database.
11. MMEX 1.8.1 will now create Mydb.emb on disk and display it - confirm by checking the very top line of the MMEX display page

I stress this worked for me on a Win 11 setup

**Comment by valsamis-d** [comment]
Thanks for the detailed step-by-step, @grumpy235 

Unfortunately, it did NOT work for me. In Step 10 MMEX fails with the following error:

![391226955-c6fd3eaa-cd48-4900-8ca3-92fafbf4c1aa](https://github.com/user-attachments/assets/29d803f9-8c3a-4e12-a510-e64dfee8d02d)

**Comment by guanlisheng** [comment]
hi @valsamis-d, you are almost there, a little bit weird that step 3 failed. it was illustrated  in the [forum thread](https://forum.moneymanagerex.org/viewtopic.php?t=5569)

the detailed practice of @grumpy235 is great.

**Comment by reddyshyam** [comment]
Workarounds is fine but I feel this shouldn't be the way upgrade should happen minor or major. Hope one the devs finds a solution and we get a quick fix release.

**Comment by grumpy235** [comment]
@guanlisheng, @whalley, @n-stein  
Guys, I did a quick search for "sqlite rekeying failed" and the top hit is for wxSQLite3 and from my brief look with limited knowledg then there might be an intermittent bug in this area.
1.8.1 was delivered with wx SQLite3 at 4.9.1 (Nov '22 !!) and Ulrich has now moved wxSQLite3 onto 4.9.12 as of Oct 24
Might be worth an update ??

**Comment by vinodraut** [comment]
Said issues is on Mac too

**Comment by Eduardo3382** [comment]
And Linux (Ubuntu 24.04), using Flatpak version as well.

**Comment by NeoXTof** [comment]
> The encryption cipher for 1.8.1 was upgraded from AES-128 to AES-256 to be compatible with MMEX Android. To upgrade your DB to be compatible with the new cipher:
> 
> 1. Download [wxSQLitePlus](https://github.com/guanlisheng/wxsqliteplus/releases)
> 2. Open your desktop database file (either encrypted or plain)
> 3. File -> Encryption Key will upgrade the cipher to SQLCipher/AES256 in place.
> 
> Instructions copied from [this forum post](https://forum.moneymanagerex.org/viewtopic.php?p=24957#p24957) by @guanlisheng.

Worked for me on first try !

**Comment by reddyshyam** [comment]
> > The encryption cipher for 1.8.1 was upgraded from AES-128 to AES-256 to be compatible with MMEX Android. To upgrade your DB to be compatible with the new cipher:
> > 
> > 1. Download [wxSQLitePlus](https://github.com/guanlisheng/wxsqliteplus/releases)
> > 2. Open your desktop database file (either encrypted or plain)
> > 3. File -> Encryption Key will upgrade the cipher to SQLCipher/AES256 in place.
> > 
> > Instructions copied from [this forum post](https://forum.moneymanagerex.org/viewtopic.php?p=24957#p24957) by @guanlisheng.
> 
> Worked for me on first try !

May be your database is not encrypted?

**Comment by NeoXTof** [comment]
> > > The encryption cipher for 1.8.1 was upgraded from AES-128 to AES-256 to be compatible with MMEX Android. To upgrade your DB to be compatible with the new cipher:
> > > 
> > > 1. Download [wxSQLitePlus](https://github.com/guanlisheng/wxsqliteplus/releases)
> > > 2. Open your desktop database file (either encrypted or plain)
> > > 3. File -> Encryption Key will upgrade the cipher to SQLCipher/AES256 in place.
> > > 
> > > Instructions copied from [this forum post](https://forum.moneymanagerex.org/viewtopic.php?p=24957#p24957) by @guanlisheng.
> > 
> > 
> > Worked for me on first try !
> 
> May be your database is not encrypted?

I reopened it with wxSQLLitePlus after checking with MMEX 1.8.1, I had to input my password. And I checked the content of the file with a text editor, everything is giberish, whereas in a not encrypted DB, you can clearly see the table name in the file.

And I tried opening the DB with the Android version of MMEX, it's also asking for the password and now it's properly getting the info from the db

**Comment by gabriele-v** [comment]
@guanlisheng @n-stein @vomikan @whalley
I have the same pagesize error, I thinke we should remove 1.8.1 ASAP until a permanent fix/upgrade is found within MMEX itself.
It's not the first time that MMEX has updates with changes that requires external actions or workaround, this is not acceptable for any software. Unfortunately I don't have any more time for MMEX but I don't understand how is it possible to release a new version with such a big issue without any notice/information in the changelog/release, I can't believe that no one has tested it before release especially having such a big change in DB encryption.

**Comment by reddyshyam** [comment]
> @guanlisheng @n-stein @vomikan @whalley I have the same pagesize error, I thinke we should remove 1.8.1 ASAP until a permanent fix/upgrade is found within MMEX itself. It's not the first time that MMEX has updates with changes that requires external actions or workaround, this is not acceptable for any software. Unfortunately I don't have any more time for MMEX but I don't understand how is it possible to release a new version with such a big issue without any notice/information in the changelog/release, I can't believe that no one has tested it before release especially having such a big change in DB encryption.
> 
> By the way, wxSQLitePlus workaround is not working, it returns the same pagesize error

No offense to devs, they do not get paid and are still contributing in their free time. So we and I personally appreciate all their contributions but echo @gabriele-v's opinion of reverting back to 1.8.0 as stable version and then fix the above issue at convenience and re-release it.

**Comment by reddyshyam** [comment]
> This breaking change should be highlighted in the release notes though some threads mentioned it months ago.
> 
> For the two upgrade approaches
> 
> 1. wxSQLitePlus to re-encrypt from [this forum post](https://forum.moneymanagerex.org/viewtopic.php?p=24957#p24957)
> 2. convert to plain via v1.8.0 and re-encrypt via v1.8.1 from [this forum post](https://forum.moneymanagerex.org/viewtopic.php?p=25564#p25564)
> 
> Can you guys leave your OS version (and other basic context), regardless of whether it works?

For me,

1, Didn't work
2. I am not attempting as this is not normal upgrade procedure.

I am on Windows 11 pro 64bit latest.

**Comment by guanlisheng** [comment]
thanks, @reddyshyam, for approach 1,  did you try wxSQLitePlus v0.3.9 mentioned above?

**Comment by renato-mmex** [comment]
I would suggest that users revert to version 1.8 or not use encryption until the bug is fixed.

**Comment by gabriele-v** [comment]
> This breaking change should be highlighted in the release notes though some threads mentioned it months ago.
> 
> For the two upgrade approaches
> 1. wxSQLitePlus to re-encrypt from [this forum post](https://forum.moneymanagerex.org/viewtopic.php?p=24957#p24957)
> 2. convert to plain via v1.8.0 and re-encrypt via v1.8.1 from [this forum post](https://forum.moneymanagerex.org/viewtopic.php?p=25564#p25564)
> 
> Can you guys leave your OS version (and other basic context), regardless of whether it works?
> 

Tried both without success, same page size error. Windows 11 Professional, not sure if it's related to OS but probably more related to first time that DB has been encrypted/created and maybe wxsqlite version at that time. By the way, I had to use wxSqlitePlus 3.0.8 because new one is detected as virus and immediately deleted by Windows Defender.

**Comment by guanlisheng** [comment]
thanks for all the information.  

for those who use v1.8.0, can you try converting it to plain and re-encrypting it via v1.8.0? to determine whether it exists in v1.8.0 as well.

**Comment by guanlisheng** [comment]
can anyone share a mini database to reproduce the issue?

**Comment by gabriele-v** [comment]
I've sent you an email with a demo DB with the pagesize issue

**Comment by guanlisheng** [comment]
here is the instruct to update the legacy page size (1024) to the default page size (4096).

run the following 
```
PRAGMA page_size = 4096;
VACUUM INTO '/yourpath/yourdatabasefullname.mmb';
```

![image](https://github.com/user-attachments/assets/133c6503-74f5-4c1b-b9a9-c64899198041)

**Comment by valsamis-d** [comment]
The above worked like a charm, @guanlisheng ! Thank you!!

**Comment by gabriele-v** [comment]
Confirmed working, I've tried something similar but the trick is to VACUUM in another DB and not the same.

**Comment by micoci2** [comment]
Thank you for all these explanations.

As we know now that the problem comes from the page size, will a patch be integrated in a future version of MMEX ?

Or do we have to use an SQLite Client on our own ? In that case, for macOS users, what SQLite Client do you recommend ? (as wxSqlitePlus doesn't seem to be available on that OS)

Thank you for your great work !

**Comment by freiru** [comment]
I have exactly the same problem with macOS 15.1.1. I could only solve the problem by downgrading to version 1.8.0. But deleting version 1.8.1 was first not successful. I had to use AppCleaner. I am now hesitant to upgrade to 1.8.1 since I will loose all my data. I am only a user and not familiar with any programming and definitely not doing anything with SQLite. Hope the bug will be fixed soon so I can upgrade.

**Comment by guanlisheng** [comment]
hi @micoci2 , mac osx can use built SQLite3 from its term

**Comment by paulcroft** [comment]
Good afternoon, 

I, too, am a user of emb files which I cannot open in v.1.8.1 and agree with reddyshyam.  To my mind the only logical way to resolve this issue is for the devs to fix this at source and provide another update (v.1.8.1.a?).  

I imagine I was not the only one who had a 'moment' when they couldn't open vital databases containing years of inputs and were mightily relieved to discover that the issue could at least be resolved, for now, by the simple matter of downgrading.

Paul

**Comment by guanlisheng** [comment]
My 2 cents, 

Utilizing external and/or internal tools to perform data upgrades is practical, particularly for this publicly SQLite3 data format, which at most requires a few clicks, copy, and paste.  Many of the guys have achieved the goal already.

**Comment by n-stein** [comment]
Just a small code fix. If opening the db with SQLCipher (AES256) fails, it attempts to open with the legacy AES128 cipher. If that succeeds, it prompts to auto-rekey to the new cipher:

![image](https://github.com/user-attachments/assets/72b99bce-1a28-483a-84b3-fa448ebd8b01)

**Comment by gabriele-v** [comment]
Do you want me to tested it with a DB that has the pagesize issue? I suspect it won't work

**Comment by n-stein** [comment]
> Do you want me to tested it with a DB that has the pagesize issue? I suspect it won't work

Indeed it fails if the page size is 1024. Let me work on that.

**Comment by n-stein** [comment]
Now it works with the 1024 page size

**Comment by whalley** [comment]
> Now it works with the 1024 page size.
> 
> FYI - There is a huge backlog of builds in the moneymanagerex AppVeyor pipeline, so it likely won't be available there for quite a while. You can test it out in the next 20 mins or so with my AppVeyor build here https://ci.appveyor.com/project/n-stein/moneymanagerex/builds/51095265.

Great work, appreciated. Must admit I was not aware of the compatibility issues here as I tend to use file level encryption for my DB, which works for me.

**Comment by reddyshyam** [comment]
> > Now it works with the 1024 page size.
> > FYI - There is a huge backlog of builds in the moneymanagerex AppVeyor pipeline, so it likely won't be available there for quite a while. You can test it out in the next 20 mins or so with my AppVeyor build here https://ci.appveyor.com/project/n-stein/moneymanagerex/builds/51095265.
> 
> Great work, appreciated. Must admit I was not aware of the compatibility issues here as I tend to use file level encryption for my DB, which works for me.

Worked like a charm! Great fix there! 👍👏🙏

**Comment by TieumTieum** [comment]
> > The encryption cipher for 1.8.1 was upgraded from AES-128 to AES-256 to be compatible with MMEX Android. To upgrade your DB to be compatible with the new cipher:
> > 
> > 1. Download [wxSQLitePlus](https://github.com/guanlisheng/wxsqliteplus/releases)
> > 2. Open your desktop database file (either encrypted or plain)
> > 3. File -> Encryption Key will upgrade the cipher to SQLCipher/AES256 in place.
> > 
> > Instructions copied from [this forum post](https://forum.moneymanagerex.org/viewtopic.php?p=24957#p24957) by @guanlisheng.
> 
> Worked for me on first try !

I am getting the error that has already been reported on WIn11 as well:
![image](https://github.com/user-attachments/assets/a277ebdf-05e2-4212-a1b0-4946b4ebbf3b)

I guess I'll stick with 1.8.0 until a robust solution is found

**Comment by TieumTieum** [comment]
> PRAGMA page_size = 4096;
> VACUUM;

This does not work for me

Error 1 in : 
VACUUM;
1 : SQL logic error[1]: cannot VACUUM from within a transaction

**Comment by reddyshyam** [comment]
> https://ci.appveyor.com/project/n-stein/moneymanagerex/builds/51095265.

Did you try [this](https://github.com/moneymanagerex/moneymanagerex/issues/6971#issuecomment-2512799807).

**Comment by reddyshyam** [comment]
> > PRAGMA page_size = 4096;
> > VACUUM;
> 
> This does not work for me
> 
> Error 1 in : VACUUM; 1 : SQL logic error[1]: cannot VACUUM from within a transaction


Did you try [this](https://github.com/moneymanagerex/moneymanagerex/issues/6971#issuecomment-2512799807).

**Comment by merlinz01** [comment]
wxSqlite method worked perfectly for me.

**Comment by hpsxrb** [comment]
> Estou enfrentando o mesmo problema.
> 
> Pensei que seria possível salvar um DB não criptografado em 1.8.0 e criptografá-lo novamente em 1.8.1. No entanto, tentar criptografar o DB em 1.8.1 falha com a seguinte mensagem de erro:
> 
> `SQL logic error[1]: Rekeying failed. Pagesize cannot be changed for an encrypted database.`
> 
> ![erro-lógico-sql](https://private-user-images.githubusercontent.com/1184162/391185395-56638b46-ef0d-4618-a8aa-d6213066f6cd.PNG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQxMDc3NTUsIm5iZiI6MTczNDEwNzQ1NSwicGF0aCI6Ii8xMTg0MTYyLzM5MTE4NTM5NS01NjYzOGI0Ni1lZjBkLTQ2MTgtYThhYS1kNjIxMzA2NmY2Y2QuUE5HP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTNUMTYzMDU1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NzlmYzIwNTk3ZDA3M2E1ZWZlNGI5MzZjN2E5ODIyYTliODg5YjQ0YWU5OGQ4N2JiOGYwNjEzN2FjMzRmNzYwMiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.40D9AGmSLZ-fbMinaa3LMcqVEnIH2TiQ_Zgv_dPgrIY)

me too

**Comment by hpsxrb** [comment]
I also have the problem presented here and I have tried all the solutions presented and they give the errors mentioned.

**Comment by n-stein** [comment]
@hpsxrb Have you tried the 1.9.0 beta?

**Comment by hpsxrb** [comment]
No

**Comment by hpsxrb** [comment]
Tnanks

**Comment by hpsxrb** [comment]
> @hpsxrbVocê já experimentou a versão beta 1.9.0?

It worked, thank you very much, problem solved with the development version (1.9.0 Beta.1)!
